### DISCUSSION TITLE
Subtracting Unit values behaves differently between 'in' and 'inch'

### DISCUSSION THREAD

**Discussion by radianrad** [josdejong/mathjs#3031] [Q&A]
This could be something on my end, but I'm noticing that 'in' is not always being processed as 'inch' in arithmetic, and instead is attempting the 'to' function. Seems to only occur with negative values or subtraction.

Works:
https://math.rad.cm/?title=&precision=6&expression=-3+inch+-+4+inch

Does not work:
https://math.rad.cm/?title=&precision=6&expression=-3+in+-+4+in

**Comment by radianrad** [comment]
More interesting cases:

Works:
https://math.rad.cm/?title=&precision=6&expression=0+mm+-+3+in+-+4+in
https://math.rad.cm/?title=&precision=6&expression=-+3+mm+-+4+in

Does not work:
https://math.rad.cm/?title=&precision=6&expression=-+3+in+-+4+in

**Comment by josdejong** [comment]
There is indeed a naming conflict between `in` meaning `inch`, or meaning `to` (converting a unit). 

The best way to circumvent this naming conflict is to not use `in`, and use `inch` and `to` instead 😄 

Concretely, I see one issue that may be improved upon, handling a negative value:

```
3in + 4in # 7 inch
-3in + 4in # error
```

We can do that by adding another condition in the expression parser here:

https://github.com/josdejong/mathjs/blob/c35a8019eefcbfa1ce5f220fe02e385007c2954f/src/expression/parse.js#L1062

Making it recon with a negative values too: 

```js
(state.token === 'in' && isConstantNode(node)) || 
(state.token === 'in' && isOperatorNode(node) && node.fn === 'unaryMinus' && isConstantNode(node.args[0])) ||
```

I have to think though though if this really does address this concern and whether it doesn't introduce new issues. Anyone any thoughts in this regard?

**Reply by radianrad** [reply]
Thanks! I had a feeling it might be that. No issues using inch for now. Not an option at this point for backwards compatibility, but 'to' alone could probably cover the conversion instead of 'to' and 'in'.

**Reply by dpnova** [reply]
hey @josdejong - in our system we serialise the results of calculations so we can restore them later and potentially do things like simplify or change unit system etc.

It seems that the result <> parse/eval loop is not bidirectional because we've seen something like `83.99999999999999 cm in^2` as the result for a calculation because we aren't strictly validating input formulae yet. 

In the above situation we can parse the JSON object with the reviver, but then trying to evaluate it errors out. `console.log(math.parse('83.99999999999999 cm in^2'))`

We have a few proposed solutions in front of us (including using unitmath for parsing calculation results), but I thought I'd check that we're not missing something obvious here?

The most bulletproof serialization i've found in this case is to format with explicit mult and parens - then parse that later... but that seems like something we should be able to do with the JSON serialization via the replacer?

**Reply by dpnova** [reply]
FTR here's the current version of the pnpm patch I'm using to sidestep the issue... if there's a way to make this a config for maths to disable at runtime I'd be happy to PR
[mathjs.patch](https://github.com/user-attachments/files/22938807/mathjs.patch)

**Reply by gwhitney** [reply]
Well, it would seem to me to be a bug if mathjs can't parse its own input. The two easy fixes (i.e., that avoid complexifying the parser even further to do a better job at correctly determining whether a given `in` is a unit or a conversion operator) are to

1. drop `in` as a conversion operator in favor of just `to`, of which it is just a synonym anyway, as in the above patch, or (less drastically)
2. just make sure mathjs never outputs `in` as a unit, e.g. always uses `inch` for output, even though it sometimes understands `in` as a unit when reading expressions.

I don't personally care much either way, but (1) is a breaking change and (2) is not, so that would typically be the way we would go. @josdejong should we file this as a bug? If so, let us know what strategy for resolution you prefer.

**Reply by dpnova** [reply]
FWIW I think there is a third option I noted above which is to make serialized output safer to parse without forcing `in to inches` - the output of `format` with explicit mult and parens seems to be a lot more parseable later - it's just you can't really get the output of format in a serialized object without updating the json after the fact or something

**Reply by dpnova** [reply]
i'd also entertain the idea that something like the output I gave is the result of a calculation - it's formatted with fairly aestheic options, which may compromise parsability. The question in my mind is whether mathjs does, or even should make a distinction between expression input syntax parsability vs result parsability. Maybe it's too fine a hair to split :)

**Reply by gwhitney** [reply]
If you do some calculation, and get a result from mathjs, and do a roundtrip to JSON and back to a mathjs entity, you should get the same entity. If that's not happening, it is a bug (and it would help if you could provide some minimal but complete example where this doesn't work, thanks).

**Reply by dpnova** [reply]
In our example, we allow the user to take a value calculated elsewhere and use it in another expression - so we are parsing the formatted result, as per this example:

```ts
console.log(math.parse(JSON.parse(JSON.stringify(math.evaluate('t=9cm * 7in^2'), math.replacer), math.reviver).format()));
```

**Reply by gwhitney** [reply]
Gotcha. And if you take out the JSON round trip, so you are just parsing the format of the result of the math.evaluate, does the problem go away?

**Reply by dpnova** [reply]
with the json serialisation the value ends up being this -> `{"mathjs":"Unit","value":63,"unit":"cm in^2","fixPrefix":false}`

whereas `console.log(math.evaluate('t=9cm * 7in^2').format());` this outputs: `24.803149606299215 in^3` (which imho seems much more correct assuming you'd rather have results in one unit system)

**Reply by gwhitney** [reply]
Sounds like some Unit object flag to do with simplifying units is not getting set to what it was originally as a result of the JSON roundtrip. @josdejong, that seems like a bug to me, that some Unit information is getting lost in the JSON roundtrip. Should we file that point as a bug? (It would either be instead of or in addition to the above question of not being able to parse '83.99999999999999 cm in^2'.)

**Reply by josdejong** [reply]
Yes, see #3562
